sources/neko_ext/mask_ops.f90

```diff
@@ -0,0 +1,136 @@
+! Copyright (c) 2024, The Neko Authors
+! All rights reserved.
+!
+! Redistribution and use in source and binary forms, with or without
+! modification, are permitted provided that the following conditions
+! are met:
+!
+!   * Redistributions of source code must retain the above copyright
+!     notice, this list of conditions and the following disclaimer.
+!
+!   * Redistributions in binary form must reproduce the above
+!     copyright notice, this list of conditions and the following
+!     disclaimer in the documentation and/or other materials provided
+!     with the distribution.
+!
+!   * Neither the name of the authors nor the names of its
+!     contributors may be used to endorse or promote products derived
+!     from this software without specific prior written permission.
+!
+! THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+! "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+! LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+! FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+! COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+! INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+! BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+! LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+! LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+! POSSIBILITY OF SUCH DAMAGE.
+!
+!> Some common Masking operations we may need
+!! TODO
+!! I know I got outvoted on defining a mask as a field of booleans...
+!! You were right, `point_zones` are convenient, but I don't like that the
+!! `invert` is defined on initialization (maybe I'm misunderstanding)
+!! So we can chose either the interior or the exterior at the beginning.
+!!
+!! There's probably a smarter way of doing this, but if you want to zero the
+!! interior vs the exterior, one will involve some awkward memory shifting.
```

Comment:
If you want to do things in two different masks you should create two masks.
So create one for the interior if you need the interior, and one for the
exterior if you need that.

Reply:
Would we now not be in the position that we're 4x heavier than a field of booleans... since combined we're effectively storing the same number of GLL pts, but now with integers?

Reply:
What is it you want to do to everything outside the mask?

Reply:
> What is it you want to do to everything outside the mask?

Take for instance, an optimization domain defined on $\Omega_{O} \subset \Omega$.
Then the design variables live on the interior, so calculating volumes etc require the inside.
But to "mask" the design requires forcing everything that's outside of $\Omega_{O}$ to 0 or 1, depending on if we want everything outside of $\Omega_{O}$ to be solid or fluid

Reply:
Oh, you intend to use these "masks" to define the design itself and not just the feasible domain of our design?

Reply:
> Oh, you intend to use these "masks" to define the design itself and not just the feasible domain of our design?

No I mean feasible domain of our design. I'm saying if you take a boolean subtraction between $\Omega$ and $\Omega_{O}$ So everything in the computational domain that is not in the optimization domain. That still has to be either solid or fluid. Which is why you need to outside.

Reply:
Lets discuss it tomorrow.

From what I see we don't need to explicitly do anything on just the exterior.

Reply:
Sure!
My morning to totally filled with meetings and seminar, so in the arvo?

Maybe a final thing to be thinking about, in terms of explicitly doing things on the interior vs the exterior, is where we define these variables. For instance, I agree that $\rho(\mathbf{x})$ should be defined with $\mathbf{x} \in \Omega_O$, but when we start mapping etc, something like $\chi (\mathbf{x})$ I'm saying $\mathbf{x} \in \Omega$.

It gets worse when we think about the PDE filters. $\tilde{\rho}(\mathbf{x})$, do we assert this is applicable for $\mathbf{x} \in \Omega_O$ or $\mathbf{x} \in \Omega$. We could interpreter this different way  (I drew a little picture to explain)

In circumstances such as these, "masking" a field isn't applying an operation only to the GLL pts inside the mask. It's applying an operation to the entire field, and only retaining what's inside the mask.

See you tomorrow!

![PXL_20241016_144904151](https://github.com/user-attachments/assets/24df4392-093f-4a13-8539-c3e095dd00f5)

Reply:
> > well the majority of mask applications end up being used to force some GLL pts to zero. So the applications in the loops could be `fld(i) = real(mask(i),kind=rp)*fld(i)`
> 
> Why would you update GLL points in the first place that you don't want to update. In any case, you are probably holding your information in a temporary array, so you can either set everything outside to zero, or just only copy everything inside to another variable.

well yeh, I mean that's exactly what `mask_exterior_const` is doing.